content/blog/2022/2022-03-28-complete-guide-to-errors-in-express.md

```diff
@@ -0,0 +1,559 @@
+
+We can run our application with the `node` command:
+
+```shell
+node js/index.js
+```
+This will start a server that will listen for requests in port `3000`.  
+
+We have also defined a server application in a file: `js/server.js` which we can run to simulate an external service. We can run the server application with the command:
+
+```js
+node js/server.js
+```
+This will start the server application on port `3001` where we can access a REST API on a URL: `http://localhost:3001/products`. We will call this service in some of our examples to test errors related to an external API call.
+
+The application in `index.js` does not contain any error handling code as yet. Node.js applications crash when they encounter unhandled exceptions. So we will next add code to this application for simulating different error conditions and handling them in the subsequent sections.
+
+## Handling Errors in Route Handler Functions
+The simplest way of handling errors in Express applications is by putting the error handling logic in the individual route handler functions. We can either check for specific error conditions or use a `try-catch` block for intercepting the error condition before invoking the logic for handling the error. 
+
+Examples of error handling logic could be logging the error stack to a log file or returning a helpful error response. 
+
+An example of a error handling in a route handler function is shown here: 
+
+```js
+const express = require('express')
+const app = express()
+
+app.use('/products', express.json({ limit: 100 }))
+
+// handle post request for path /products
+app.post('/products', (request, response) => {
+  const name = request.body.name                
+  ...
+  ...
+
+  // Check for error condition
+  if(name == null){
+    // Error handling logic: log the error
+    console.log("input error")
+
+    // Error handling logic: return error response
+    response
+      .status(500)
+      .json({ message: "Mandatory field: name is missing. " })
+  }else{
+    // continue with normal processing             
+    const productCreationResponse = { result: "success"}
+
+    // return success response
+    response.json(productCreationResponse)
+  }
+})
+```
+Here we are checking for the error condition by checking for the presence of a mandatory input in the request payload and returning the error as an HTTP error response with error code 500 and an error message as part of the error handling logic.
+
+Here is one more example of handling error using a `try-catch` block:
+
+```js
+const express = require('express')
+const axios = require("axios")
+const app = express()
+
+app.get('/products', async (request, response)=>{
+  try{
+    const apiResponse = await axios.get("http://localhost:3001/products")
+
+    const jsonResponse = apiResponse.data
+    console.log("response "+jsonResponse)
+    
+    response.send(jsonResponse)
+  }catch(error){ // intercept the error in catch block
```

Comment:
Please check all code examples for proper formatting. There are some without spaces between spaces like here.